### PR TITLE
Import SpiderMonkey Temporal tests

### PR DESCRIPTION
**Pull request by anba**
Temporal tests written for the SpiderMonkey implementation. Mostly
covers edge cases around mathematical operations and regression tests
for reported spec bugs.

### CODE REVIEW

**Comment by Ms2ger** [comment]
Linter fails - I think you're missing the year in your copyright headers

File: test/built-ins/Temporal/Duration/compare/precision-exact-mathematical-values-1.js
Comment (ptomato) [inline test/built-ins/Temporal/Duration/compare/precision-exact-mathematical-values-1.js:23]:
```suggestion
  10. Let ns2 be ! TotalDurationNanoseconds(days2, two.[[Hours]], two.[[Minutes]], two.[[Seconds]],
     two.[[Milliseconds]], two.[[Microseconds]], two.[[Nanoseconds]], shift2).
  11. If ns1 > ns2, return 1𝔽.
  12. If ns1 < ns2, return -1𝔽.
  13. Return +0𝔽.
```
(dedenting by one position seems to mess up the YAML parser)

File: test/built-ins/Temporal/Duration/compare/precision-exact-mathematical-values-2.js
Comment (ptomato) [inline test/built-ins/Temporal/Duration/compare/precision-exact-mathematical-values-2.js:23]: Same here.

File: test/intl402/Temporal/TimeZone/non-canonical-africa.js
Comment (ptomato) [inline test/intl402/Temporal/TimeZone/non-canonical-africa.js:1]: Out of curiosity what's the reason the non-canonical time zone name tests can't be in one file?

**Review by ptomato — COMMENTED** [review]
Thanks. I expect to be able to look at these in detail next week, although I need to look at https://github.com/tc39/test262/pull/3523 first, so it will depend on when I get through with that one.

I've given a quick try to running these and have left some comments based on that.

File: test/built-ins/Temporal/PlainTime/prototype/add/precision-exact-mathematical-values-2.js
Comment (anba) [reply test/built-ins/Temporal/PlainTime/prototype/add/precision-exact-mathematical-values-2.js]: Actually I've removed this line, because it was a leftover from the original test and the assertion is superseded by `TemporalHelpers.assertPlainTime`.

File: test/built-ins/Temporal/Duration/prototype/add/days-is-number-max-value-with-zoneddatetime.js
Comment (anba) [reply test/built-ins/Temporal/Duration/prototype/add/days-is-number-max-value-with-zoneddatetime.js:5]: Added `subtract` tests for all `add` tests under "Temporal/Duration" and "Temporal/PlainTime".

File: test/intl402/Temporal/TimeZone/non-canonical-africa.js
Comment (ptomato) [reply test/intl402/Temporal/TimeZone/non-canonical-africa.js:1]: Wouldn't all the files have to be skipped anyway if CLDR canonicalization was used?

We could also assert that the ID is equal to either the expected value for CLDR or IANA, but in practice that probably doesn't tell us anything more than your second option (only testing that the non-canonical name is allowed.)

File: test/intl402/Temporal/TimeZone/non-canonical-antarctica.js
Comment (anba) [reply test/intl402/Temporal/TimeZone/non-canonical-antarctica.js:12]: ECMA-402 assumes that all links are in `backward`, which is a [known error](https://tc39.es/archives/bugzilla/1892/). For example following the spec by the letter means that `"Etc/GMT"` is canonicalised to `"UTC"` (per [CanonicalizeTimeZoneName](https://tc39.es/ecma402/#sec-canonicalizetimezonename), step 3), whereas `"GMT"` shouldn't be canonicalised, because the `GMT` &rarr; `Etc/GMT` link is in [etcetera](https://github.com/eggert/tz/blob/bf8aa9414b0de5fc9df5f25fa12d07451510e926/etcetera#L26). 

https://github.com/tc39/ecma402/issues/121#issuecomment-323929572 proposes to follow all links in the main time zone files (i.e. excluding `backzone`), but implementers are unlikely to completely ignore `backzone` because of the possible negative cultural/political effects mentioned in https://github.com/tc39/ecma402/issues/272#issuecomment-423928522.

File: test/intl402/Temporal/TimeZone/non-canonical-africa.js
Comment (anba) [reply test/intl402/Temporal/TimeZone/non-canonical-africa.js:1]: > Wouldn't all the files have to be skipped anyway if CLDR canonicalization was used?

Not necessarily all files, for example all links in "non-canonical-brazil.js" are exactly the same in IANA and CLDR.

File: test/intl402/Temporal/TimeZone/non-canonical-antarctica.js
Comment (ptomato) [reply test/intl402/Temporal/TimeZone/non-canonical-antarctica.js:12]: If the links for Antarctica/South_Pole in `backward` and `backzone` conflict, then which one should take precedence in your opinion?

(If this is an existing web reality issue then it seems better to me that we only check that the link names _can_ be canonicalized, and not what they are canonicalized to...)

File: test/intl402/Temporal/TimeZone/prototype/getNextTransition/subtract-second-and-nanosecond-from-last-transition.js
Comment (ptomato) [inline test/intl402/Temporal/TimeZone/prototype/getNextTransition/subtract-second-and-nanosecond-from-last-transition.js:20]: FYI, we are usually avoiding using `toString()` to make assertions about Temporal objects, though here I think it is justified for easy comparison to the TZDB excerpt above.

File: test/intl402/Temporal/TimeZone/prototype/getNextTransition/transition-at-instant-boundaries.js
Comment (ptomato) [inline test/intl402/Temporal/TimeZone/prototype/getNextTransition/transition-at-instant-boundaries.js:28]: This is no longer the case, I think?

File: test/intl402/Temporal/TimeZone/prototype/getOffsetNanosecondsFor/nanoseconds-subtracted-or-added-at-dst-transition.js
Comment (ptomato) [inline test/intl402/Temporal/TimeZone/prototype/getOffsetNanosecondsFor/nanoseconds-subtracted-or-added-at-dst-transition.js:14]: Small detail, but isn't this the wrong line for the data used in the test below? We need the one about the [last Sunday in April](https://github.com/eggert/tz/blob/main/northamerica#L559)

File: test/intl402/Temporal/TimeZone/prototype/getPreviousTransition/transition-at-instant-boundaries.js
Comment (ptomato) [inline test/intl402/Temporal/TimeZone/prototype/getPreviousTransition/transition-at-instant-boundaries.js:28]: Same comment as in the corresponding getNextTransition test

**Review by ptomato — COMMENTED** [review]
I didn't go through everything, but I did verify a big chunk of these — would you prefer to keep everything together in one PR or would you like me to commit the ones that I've done already?

File: test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-too-large.js
Comment (anba) [reply test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-too-large.js:85]: Yes.

File: test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-zero-or-negative.js
Comment (anba) [reply test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-zero-or-negative.js:67]: Yes.

File: test/intl402/Temporal/TimeZone/prototype/getNextTransition/transition-at-instant-boundaries.js
Comment (anba) [reply test/intl402/Temporal/TimeZone/prototype/getNextTransition/transition-at-instant-boundaries.js:28]: <https://github.com/tc39/proposal-temporal/pull/2351> will ensure no error is thrown.

File: test/intl402/Temporal/TimeZone/prototype/getOffsetNanosecondsFor/nanoseconds-subtracted-or-added-at-dst-transition.js
Comment (anba) [reply test/intl402/Temporal/TimeZone/prototype/getOffsetNanosecondsFor/nanoseconds-subtracted-or-added-at-dst-transition.js:14]: Thanks for catching this error!

**Comment by anba** [comment]
> I didn't go through everything, but I did verify a big chunk of these — would you prefer to keep everything together in one PR or would you like me to commit the ones that I've done already?

It's all the same to me. Let's do whatever works best for you.

File: test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-too-large.js
Comment (ptomato) [reply test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-too-large.js:85]: Assuming we get consensus on those normative PRs this week, I'd rather just wait a few days and commit the updated behaviour, so let's adjust these (here and below) before merging.

**Comment by ptomato** [comment]
> > I didn't go through everything, but I did verify a big chunk of these — would you prefer to keep everything together in one PR or would you like me to commit the ones that I've done already?
> 
> It's all the same to me. Let's do whatever works best for you.

I've done this in #3606.

File: test/intl402/Temporal/TimeZone/prototype/getNextTransition/transition-at-instant-boundaries.js
Comment (ptomato) [reply test/intl402/Temporal/TimeZone/prototype/getNextTransition/transition-at-instant-boundaries.js:28]: I made that change and moved this into #3610.

File: test/intl402/Temporal/TimeZone/prototype/getPreviousTransition/transition-at-instant-boundaries.js
Comment (ptomato) [reply test/intl402/Temporal/TimeZone/prototype/getPreviousTransition/transition-at-instant-boundaries.js:28]: I made that change and moved this into #3610.

**Comment by ptomato** [comment]
My plan for following up:
1. Separate out the tests that are affected by the recently adopted Temporal normative changes, update them to reflect the changes, and merge them at the same time as the corresponding normative PRs
2. Then, review everything else that's not number precision issues
3. Save the most difficult work for last :stuck_out_tongue_closed_eyes: 

@anba, if you have time to do the updates for (1) I'd appreciate it, otherwise I'll get around to it sometime this week or next.

File: test/intl402/Temporal/TimeZone/non-canonical-antarctica.js
Comment (anba) [reply test/intl402/Temporal/TimeZone/non-canonical-antarctica.js]: I've updated the tests to only check that link names are accepted.

File: test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-too-large.js
Comment (anba) [reply test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-too-large.js]: #3617 updates this test to reject negative day lengths.

File: test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-zero-or-negative.js
Comment (anba) [reply test/built-ins/Temporal/ZonedDateTime/prototype/round/smallest-unit-day-daylength-zero-or-negative.js]: #3617 updates this test to reject negative day lengths.

**Comment by anba** [comment]
> @anba, if you have time to do the updates for (1) I'd appreciate it, otherwise I'll get around to it sometime this week or next.

I've added the following four commits:
- The first one updates a test which uses negative day lengths.
- The other three commits address review comments from your first review.

**Review by ptomato — COMMENTED** [review]
Thanks, I've merged another batch in, and rebased this one on top.

File: test/intl402/Temporal/TimeZone/non-canonical-utc.js
Comment (anba) [reply test/intl402/Temporal/TimeZone/non-canonical-utc.js:15]: I think it's okay to test `Etc/GMT0` in both files, because it's both a legacy name and also another alias for `UTC`.

File: test/built-ins/Temporal/Duration/compare/precision-exact-mathematical-values-1.js
Comment (ptomato) [inline test/built-ins/Temporal/Duration/compare/precision-exact-mathematical-values-1.js:58]: For future reference, better to add a message here (and elsewhere) like `new Test262Error("dateAdd called with unexpected value")` to aid in debugging implementations. Don't worry about fixing it up, I can add this before I push it.

**Comment by ptomato** [comment]
Sorry it took so long. I've reviewed another batch of these and they're ready to land in https://github.com/tc39/test262/pull/3677. Once that is merged, I'll rebase this again and review the remaining 5 tests. I expect to be able to get to that early next week.

**Review by ptomato — APPROVED** [review]
Thanks very much for the thorough tests, as well as for bearing with me for so long while I found time to work through the arithmetic ... these are unbelievably helpful.

**Review by ptomato — APPROVED** [review]
Approving again after force-push

**Pull request merged**
